packages/langium/src/grammar/grammar-util.ts

```diff
@@ -308,7 +308,7 @@ function withCardinality(regex: string, cardinality?: string, wrap = false): str
 
 export function getTypeName(type: ast.AbstractType | ast.InferredType): string {
     if (ast.isParserRule(type)) {
-        return getExplicitRuleType(type) ?? type.name;
+        return getExplicitRuleTypeName(type) ?? type.name;
```

Comment:
Referencing the comment below, here would be a great spot to pull such a check up to. See below for details.